### ISSUE TITLE
Why is "a"+"b" mutated to "a"-"b" and compiles?

### ISSUE THREAD

**Issue by cs-util** [stryker-mutator/stryker-net#304]
How comes that the mutation report has lines like this:
`[Survived] Binary expression mutation on line 26: '"" + toString(x)' ==> '"" - toString(x)'`

Mutating `"a"+"b"` to `"a"-"b"` is not valid C# code if I am not totally wrong?  So why is this a valid mutation that even passes the execution of my tests?

**Comment by richardwerkman** [comment]
Could you please provide the original code as well? 

Some background info: Stryker mutates all code it finds. Then compiles and rollbacks all mutations that caused compile errors. If this mutation somehow passed the rollback process it probably compiled correctly.

**Comment by cs-util** [comment]
Here is one of the many lines in the report that looks suspicious to me since they can't produce compiling code:

`[Survived] Binary expression mutation on line 18: 'utcString.Substring(0, "UTC", false) + "GMT"' ==> 'utcString.Substring(0, "UTC", false) - "GMT"'`

Here is the code with the line and I did the mutation manually to check if it really does not compile:
![image](https://user-images.githubusercontent.com/45512166/51011844-7b37b880-155a-11e9-92a0-7b5097715969.png)

**Comment by cs-util** [comment]
I also found other lines in the report that are very strange:

For example, I have an extension method in a :
```
public static bool InvokeIfNotNull(this Action a) { 
        if (a == null) { return false; } else { a.Invoke(); return true; } 
}
```
![image](https://user-images.githubusercontent.com/45512166/51014437-efc52400-1567-11e9-9493-a5c416c9a35e.png)

And a unit test for it:

```
[Fact]
public void DelegateExtensionTests() {
    Action a = null;
    Assert.False(a.InvokeIfNotNull());
    var wasCalled = false;
    a = () => { wasCalled = true; };
    Assert.True(a.InvokeIfNotNull());
    Assert.True(wasCalled);
}
```

But Stryker reports:
```
[Survived] Binary expression mutation on line 8: 'a == null' ==> 'a != null'
[Survived] Boolean mutation on line 8: 'false' ==> 'true'
[Killed] Boolean mutation on line 8: 'true' ==> 'false'
```

Which can't be based on the Unit test correct?

**Comment by richardwerkman** [comment]
@cs-util I've tried to recreate the compiling "string" - "string" issue but couldn't achieve your output. 

``` csharp
string test = "";
if (test.Contains("UTC"))
{
    test = test.Substring(0, 1) + "GMT";
}
```
Results only in 3 mutations for me:
```
--- Program.cs [0/3 (0.00 %)]
[Survived] String mutation on line 17: '""' ==> '"Stryker was here!"'
[Survived] String mutation on line 18: '"UTC"' ==> '""'
[Survived] String mutation on line 20: '"GMT"' ==> '""'
```
Am I missing something?

**Comment by richardwerkman** [comment]
Also, I tried running your extension method with your test and this resulted in:
```
--- ExampleExtension.cs [3/3 (100.00 %)]
[Killed] Binary expression mutation on line 11: 'a == null' ==> 'a != null'
[Killed] Boolean mutation on line 11: 'false' ==> 'true'
[Killed] Boolean mutation on line 11: 'true' ==> 'false'
```
What version of Stryker are you running? Version 0.7.0 worked fine for me.

Also your .Net Core runtime could be of importance. 

```
dotnet --version
2.1.600-preview-009426
```
I hope this helps!

**Comment by cs-util** [comment]
I am using Stryker version 0.7.0 as well and .Net version 2.2.101

I think I found the problem, my target framework was set to "netstandard2.0" and I switched now to "netcoreapp2.1" as its set in the integration tests project. 

In case that's easily possible, I think letting Striker fail if a non-supported target framework is set would be a very helpful feature since currently, it does behave kind of random for unsupported target frameworks and I am pretty sure I will not be the last person opening a ticket for this in the next months ;)

**Comment by cs-util** [comment]
With target framework netstandard2.0 the results looked like this: 
```
Tests progress | ██████████ | 779 / 779 | 100 % |
Killed : 290
Survived: 489
Timeout : 0
```

Now with target framework netcoreapp2.1 they look like this:
```
Tests progress | ██████████ | 779 / 779 | 100 % |
Killed : 686
Survived: 93
Timeout : 0
```

But there are still mutations that if I do them manually cause the tests to fail but the report says they survived:
```
--------------- DelegateExtensions.cs [16/21 (76.19 %)]
[Survived] Binary expression mutation on line 8: 'a == null' ==> 'a != null'
[Killed] Boolean mutation on line 8: 'false' ==> 'true'
[Killed] Boolean mutation on line 8: 'true' ==> 'false'
```

And there are also still mutations that produce non-compiling code, for example
`[Survived] Binary expression mutation on line 26: '"" + toString(x)' ==> '"" - toString(x)'`
Mutating this method:
![image](https://user-images.githubusercontent.com/45512166/51069906-83febc00-1638-11e9-8b40-4860bb8d161c.png)

**Comment by richardwerkman** [comment]
Thanks for looking into this further! I was not aware that the target projects "target framework" had this impact. I assumed Stryker would be backwards compatible since .net also is. 

Your suggestion to kill Stryker when a not supported target framework is found would not have my preference. I think Stryker should support netcoreapp1.1 and up. Also netstandard2.0 should be supported. We could start by creating an integration test for each target framework we want to support. All should give the same output. If this is not the case we should fix this.

**Comment by cs-util** [comment]
@richardwerkman Here is the repository for you to try: https://github.com/cs-util-com/cscore 
The relevant xUnit project for you to run is https://github.com/cs-util-com/cscore/tree/master/CsCore/xUnitTests

**Comment by richardwerkman** [comment]
@cs-util Did you by any chance already figure this out yourself? I now get the same output as you did: `[Survived] Binary expression mutation on line 26: '"" + toString(x)' ==> '"" - toString(x)'`
But this is because of a wrong implementation of the fix I did. I'll create a pull request hoping to fix your issue. If you could test this that would be great!

**Comment by richardwerkman** [comment]
Yes, the fix is released with 0.8.2! I hope it will solve the unexpected output issue. And good luck with your test stability :+1:
